src/services/redisSessionStorage.ts

```diff
@@ -2,52 +2,128 @@
   public set(token: string, user: User): void {
-    // Set key to hold the string value. This data is set to expire (EX) after
-    // `this.tokenDuration` seconds.
-    // @see https://redis.io/commands/set
-    this.redisClient.set(token, JSON.stringify(user), "EX", this.tokenDuration);
+    // Set a key to hold the fields value.
+    // @see https://redis.io/commands/hmset
+    this.redisClient.hmset(token, {
+      data: JSON.stringify(user),
+      timestamp: Date.now()
+    });
   }
 
   /**
    * {@inheritDoc}
    */
-  public get(token: string): Promise<Either<Error, User>> {
-    const client = this.redisClient;
-
+  public get(token: string): Promise<Either<Error, ISessionState>> {
     return new Promise(resolve => {
-      // Get the value of key.
-      // @see https://redis.io/commands/get
-      client.get(token, (err, value) => {
+      // Get the fields for this key.
+      // @see https://redis.io/commands/hmget
+      this.redisClient.hgetall(token, (err, value) => {
         if (err) {
-          resolve(left<Error, User>(new Error(err.message)));
+          resolve(left<Error, ISessionState>(new Error(err.message)));
         } else {
           if (value === null || value === undefined) {
-            resolve(left<Error, User>(new Error("Session not found")));
+            resolve(
+              left<Error, ISessionState>(new Error(SESSION_NOT_FOUND_MESSAGE))
+            );
           } else {
-            const errorOrUser = extractUserFromJson(value);
+            // check if the token has expired. We don't remove expired token
+            // because client can later refresh the session.
+            const timestamp = +value.timestamp;
+            if (timestamp + this.tokenDuration * 1000 < Date.now()) {
+              resolve(
+                left<Error, ISessionState>(new Error("Token has expired"))
+              );
+            }
 
-            // TODO: better error message.
-            errorOrUser.mapLeft(() => {
-              return "Errors in validating the user profile";
-            });
+            const errorOrUser = extractUserFromJson(value.data);
 
-            resolve(errorOrUser);
+            if (isLeft(errorOrUser)) {
+              resolve(
+                left<Error, ISessionState>(
+                  new Error("Unable to decode the user")
+                )
+              );
+            } else {
+              const user = errorOrUser.value;
+              resolve(
+                right<Error, ISessionState>({
+                  expireAt: "",
+                  expired: false,
+                  user
+                })
+              );
+            }
+          }
+        }
+      });
+    });
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public refresh(token: string): Promise<Either<Error, ISessionState>> {
+    return new Promise(resolve => {
+      // Get the fields for this key.
+      // @see https://redis.io/commands/hmget
+      this.redisClient.hgetall(token, (err, value) => {
```

Comment:
especially in this functions with nested branches/conditions, it would be very nice to have short comments at the beginning of each block explaining the condition (same for the `get` method above)